feat: add input state persistence

Tronix117 opened the pull request:
This feature allows to save and restore input text between sessions

- save input text to state.ron when window closes
- restore text when application launches
- clear state when match is selected in the result list
- add `enable_state_saving` option (`false` by default)
-  home-manager module support
- update documentation and examples

Use case : when configured with keybind like `pkill anyrun || anyrun`, results are dismissed between sessions. It's disturbing when doing calculations with data in multiple place and not directly available on the screen.

I've thoroughly testes thoses modifications without finding any bug, regression or issue.

Changes are totaly backward compatible, since a config option need to be enabled.

Feel free to give me your feedback

Kirottu commented on anyrun/src/main.rs:
```diff
@@ -178,6 +184,67 @@ struct RuntimeData {
 
+impl RuntimeData {
+    fn state_file(&self) -> String {
+        format!("{}/state.txt", self.config_dir)
```
The state file should not be in the config folder at all, it should be in somewhere like `~/.local/state/anyrun` or `~/.cache/anyrun`

NotAShelf replied:
I would appreciate if this followed the XDG spec, e.g., `$XDG_CACHE_HOME` or `$XDG_STATE_HOME` by default

Tronix117 replied:
Can I use dirs crate for this one and the config_dir ?

Kirottu commented on anyrun/src/main.rs:
```diff
@@ -178,6 +184,67 @@ struct RuntimeData {
+            .as_millis();
+        
+        let mut file = fs::File::create(self.state_file())?;
+        writeln!(file, "{}", timestamp)?;
```
I'd much prefer if the state file was a struct serialized with ron like all the other config related files, this would be more consistent and allow for further extension if necessary. Something along the lines of:
```
struct PersistentState {
    timestamp: u128,
    text: String,
}
```

Tronix117 replied:
I first did that, but changed it for a text file, to avoid unnecessary serialization/deserialisation on some simple structure like this one.
But it make sense to keep it !

Kirottu commented on anyrun/src/main.rs:
```diff
@@ -178,6 +184,67 @@ struct RuntimeData {
+        if !self.config.persist_state {
+            return Ok(());
+        }
+        fs::write(self.state_file(), "0\n")
```
If you are clearing the state you could just delete the file.

Kirottu commented on anyrun/src/main.rs:
```diff
@@ -679,11 +766,17 @@ fn activate(app: &gtk::Application, runtime_data: Rc<RefCell<RuntimeData>>) {
+            // Show initial results if state restoration is enabled or immediate results are configured
+            let should_show_results = {
+                let data = runtime_data.borrow();
+                data.config.persist_state || data.config.show_results_immediately
```
This does not account for whether or not there is a valid persisted state, as in will still show results immediately even if there is no input which may not be what the user wants. Should check if a valid, non-expired state exists.

Kirottu commented on nix/modules/home-manager.nix:
```diff
@@ -144,6 +144,18 @@ in {
         description = "Show search results immediately when Anyrun starts";
       };
 
+      persistState = mkOption {
```
@NotAShelf should have a look at the Nix parts to see if they are fine.

NotAShelf replied:
Nix side looks fine, nothing to note.

Kirottu commented on anyrun/src/main.rs:
```diff
@@ -178,6 +184,67 @@ struct RuntimeData {
+        write!(file, "{}", text)
+    }
+
+    fn load_state(&self) -> io::Result<String> {
```
If no valid state exists, should return None instead of empty string.

NotAShelf commented on anyrun/src/main.rs:
```diff
@@ -465,10 +532,21 @@ fn activate(app: &gtk::Application, runtime_data: Rc<RefCell<RuntimeData>>) {
+    // Update last_input, save state and refresh matches when text changes
     let runtime_data_clone = runtime_data.clone();
     entry.connect_changed(move |entry| {
-        refresh_matches(entry.text().to_string(), runtime_data_clone.clone())
+        let text = entry.text().to_string();
+        if let Err(e) = runtime_data_clone.borrow().save_state(&text) {
+            eprintln!("Failed to save state: {}", e);
+        }
+        refresh_matches(text, runtime_data_clone.clone());
```
From the top of my head, I think this is potentially a performance flaw as it will be writing to the state file on every text change, and that can cause I/O performance issues- especially if the user types fast (edge case?)

Perhaps a debounce mechanism if you could implement it seamlessly?

```diff
@@ -471,546 +471,550 @@
     entry.connect_changed(move |entry| {
         let text = entry.text().to_string();
-        if let Err(e) = runtime_data_clone.borrow().save_state(&text) {
-            eprintln!("Failed to save state: {}", e);
-        }
+        let runtime_data_clone = runtime_data_clone.clone();
+        glib::timeout_add_local(Duration::from_millis(500), move || {
+            if let Err(e) = runtime_data_clone.borrow().save_state(&text) {
+                eprintln!("Failed to save state: {}", e);
+            }
+            glib::Continue(false)
+        });
         refresh_matches(text, runtime_data_clone.clone());
     });
```

via [timeout_add_local](https://docs.rs/glib/latest/glib/source/fn.timeout_add_local.html)

Tronix117 replied:
I wanted to do that at first, but my though was that the i/o on that was pretty limited, like the user could never even if it tries pose a threat to i/o perf by typing quickly.
However I can do that if you think it's better

Kirottu replied:
Another option would be only saving it on exit but that would require handling for termination signals somehow.

Tronix117 commented:
Thanks for your feedbacks, I'll change that !

To avoid a little delay at launch for first display of matches, I also wanted to save matches in the file (so having the save after the refresh_matches has been done), however I can't find a way to easily serialize which plugin has generated the match...

Tronix117 commented on anyrun/src/main.rs:
```diff
@@ -874,7 +1029,7 @@ fn refresh_matches(input: String, runtime_data: Rc<RefCell<RuntimeData>>) {
         // If a plugin has requested exclusivity, respect it
         if let Some(exclusive) = &runtime_data.borrow().exclusive {
-            if plugin_view.plugin.info() == exclusive.plugin.info() {
+            if ptr::fn_addr_eq(plugin_view.plugin.info(), exclusive.plugin.info()) {
```
i fixed a warning here

Tronix117 replied:
You're right not a concern for this PR, I removed the commit!

Tronix117 commented:
So my last changes address all your feedbacks.

I used `dirs` crate to find relevant state and config folder. I changed it too for the config folder since it made more sense to handle them the same (+ it resolves #208  at the same time)

I moved the `runtime_data` initialisation block (was in `main` inside the `new` of the `RuntimeData` impl, and only keeped in `main` dependencies for the `RuntimeData`. It better respects the separation of concerns and was needed to avoid poluting `main` with `state_dir` and `state_file`.

I managed to save only when window will close, I think it is safe with anyrun since this is the main window. When it closes the program exits, and when we gracefuly kill the program, the window close event is also called.

Finaly I used a `ron` file like suggested and as always updated the docs.

Tronix117 commented:
wait before review, I found that now the saving mecanism skip some saving sometime, it seems the window close event is not so reliable, i put the pr in draft
